File: go/test/api_entities_personal_access_token_with_last_used_ip_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestApiEntitiesPersonalAccessTokenWithLastUsedIpEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ApiEntitiesPersonalAccessTokenWithLastUsedIp(nil)
		if ent == nil {
			t.Fatal("expected non-nil ApiEntitiesPersonalAccessTokenWithLastUsedIpEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := api_entities_personal_access_token_with_last_used_ipBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "api_entities_personal_access_token_with_last_used_ip." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.api_entities_personal_access_token_with_last_used_ip", setup.data)))
		var apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Data map[string]any
		if len(apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataRaw) > 0 {
			apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Data = core.ToMapAny(apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Data

		// LIST
		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Ent := client.ApiEntitiesPersonalAccessTokenWithLastUsedIp(nil)
		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Match := map[string]any{}

		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01ListResult, err := apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Ent.List(apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, apiEntitiesPersonalAccessTokenWithLastUsedIpRef01ListOk := apiEntitiesPersonalAccessTokenWithLastUsedIpRef01ListResult.([]any)
		if !apiEntitiesPersonalAccessTokenWithLastUsedIpRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", apiEntitiesPersonalAccessTokenWithLastUsedIpRef01ListResult)
		}

		// LOAD
		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01MatchDt0 := map[string]any{
			"id": apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Data["id"],
		}
		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataDt0Loaded, err := apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Ent.Load(apiEntitiesPersonalAccessTokenWithLastUsedIpRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataDt0LoadResult := core.ToMapAny(apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataDt0Loaded)
		if apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if apiEntitiesPersonalAccessTokenWithLastUsedIpRef01DataDt0LoadResult["id"] != apiEntitiesPersonalAccessTokenWithLastUsedIpRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func api_entities_personal_access_token_with_last_used_ipBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "api_entities_personal_access_token_with_last_used_ip", "ApiEntitiesPersonalAccessTokenWithLastUsedIpTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read api_entities_personal_access_token_with_last_used_ip test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse api_entities_personal_access_token_with_last_used_ip test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"api_entities_personal_access_token_with_last_used_ip01", "api_entities_personal_access_token_with_last_used_ip02", "api_entities_personal_access_token_with_last_used_ip03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/ApiEntitiesPersonalAccessTokenWithLastUsedIpEntityTest.php
<?php
declare(strict_types=1);

// ApiEntitiesPersonalAccessTokenWithLastUsedIp entity test

require_once __DIR__ . '/../gitlab_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class ApiEntitiesPersonalAccessTokenWithLastUsedIpEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GitlabSDK::test(null, null);
        $ent = $testsdk->ApiEntitiesPersonalAccessTokenWithLastUsedIp(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = api_entities_personal_access_token_with_last_used_ip_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "api_entities_personal_access_token_with_last_used_ip." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $api_entities_personal_access_token_with_last_used_ip_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.api_entities_personal_access_token_with_last_used_ip")));
        $api_entities_personal_access_token_with_last_used_ip_ref01_data = null;
        if (count($api_entities_personal_access_token_with_last_used_ip_ref01_data_raw) > 0) {
            $api_entities_personal_access_token_with_last_used_ip_ref01_data = Helpers::to_map($api_entities_personal_access_token_with_last_used_ip_ref01_data_raw[0][1]);
        }

        // LIST
        $api_entities_personal_access_token_with_last_used_ip_ref01_ent = $client->ApiEntitiesPersonalAccessTokenWithLastUsedIp(null);
        $api_entities_personal_access_token_with_last_used_ip_ref01_match = [];

        $api_entities_personal_access_token_with_last_used_ip_ref01_list_result = $api_entities_personal_access_token_with_last_used_ip_ref01_ent->list($api_entities_personal_access_token_with_last_used_ip_ref01_match, null);
        $this->assertIsArray($api_entities_personal_access_token_with_last_used_ip_ref01_list_result);

        // LOAD
        $api_entities_personal_access_token_with_last_used_ip_ref01_match_dt0 = [
            "id" => $api_entities_personal_access_token_with_last_used_ip_ref01_data["id"],
        ];
        $api_entities_personal_access_token_with_last_used_ip_ref01_data_dt0_loaded = $api_entities_personal_access_token_with_last_used_ip_ref01_ent->load($api_entities_personal_access_token_with_last_used_ip_ref01_match_dt0, null);
        $api_entities_personal_access_token_with_last_used_ip_ref01_data_dt0_load_result = Helpers::to_map($api_entities_personal_access_token_with_last_used_ip_ref01_data_dt0_loaded);
        $this->assertNotNull($api_entities_personal_access_token_with_last_used_ip_ref01_data_dt0_load_result);
        $this->assertEquals($api_entities_personal_access_token_with_last_used_ip_ref01_data_dt0_load_result["id"], $api_entities_personal_access_token_with_last_used_ip_ref01_data["id"]);

    }
}

function api_entities_personal_access_token_with_last_used_ip_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/api_entities_personal_access_token_with_last_used_ip/ApiEntitiesPersonalAccessTokenWithLastUsedIpTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GitlabSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["api_entities_personal_access_token_with_last_used_ip01", "api_entities_personal_access_token_with_last_used_ip02", "api_entities_personal_access_token_with_last_used_ip03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID" => $idmap,
        "GITLAB_TEST_LIVE" => "FALSE",
        "GITLAB_TEST_EXPLAIN" => "FALSE",
        "GITLAB_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GITLAB_TEST_API_ENTITIES_PERSONAL_ACCESS_TOKEN_WITH_LAST_USED_IP_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["GITLAB_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["GITLAB_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new GitlabSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["GITLAB_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GITLAB_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
